hail/python/hail/matrixtable.py

```diff
@@ -3276,7 +3276,18 @@ def distinct_by_col(self):
         -------
         :class:`.MatrixTable`
         """
-        return MatrixTable._from_java(self._jmt.distinctByCol())
+        index_uid = Env.get_uid()
+
+        col_key_fields = list(self.col_key)
+        t = self.key_cols_by().cols()
+
+        t = t.add_index(index_uid)
+        unique_cols = t.aggregate(
+            hl.agg.group_by(
+                hl.struct(**{f: t[f] for f in col_key_fields}), hl.agg.take(t[index_uid], 1)))
```

Comment:
wait, I think you can do this with a scan:

```
filter_cols(~hl.scan.collect_as_set(self.col_key).contains(self.col_key)
```

Reply:
While this is slick as fuck, I think it performs worse because it needs to create a set for each column of the matrix table making it O(N^2) while mine builds the dict online.  Or am I wrong about how scans are implemented?

I can change it to pull out the values inside the aggregator, that will save some serialization.

Reply:
yeah, I did realize after posting that it's quadratic in our current system. But it's totally what I want to write.

We can fix this, I think. It won't be easy and will require some C++ tricks. We should be able to expose some mutable builder for operations that read, like set membership, while still letting people map it away and copy (and do quadratic things) if they want.

Reply:
> We can fix this, I think.

I agree, 100%.  For example, we can have a physical type that corresponds to a mutable C++ set implementation.  It will be immutable in the Hail expression language, but the aggregator will be able to mutate it.